java/squeek/appleskin/client/HUDOverlayHandler.java

```diff
@@ -134,68 +151,157 @@ public static void drawSaturationOverlay(MatrixStack matrixStack, float saturati
-		int barsNeeded = endBar - startBar;
+		enableAlpha(alpha);
 		mc.getTextureManager().bindTexture(Screen.GUI_ICONS_TEXTURE);
 
-		enableAlpha(alpha);
-		for (int i = startBar; i < startBar + barsNeeded; ++i)
+		int modifiedFood = Math.min(20, foodLevel + hungerRestored);
+
+		int startFoodBars = foodLevel / 2;
+		int endFoodBars = (int)Math.ceil(modifiedFood / 2.0F);
+
+		int iconStartOffset = 16;
+		int iconSize = 9;
+
+		for (int i = startFoodBars; i < endFoodBars; ++i)
 		{
-			int idx = i * 2 + 1;
-			int x = left - i * 8 - 9;
-			int y = top;
-			int icon = 16;
-			int background = 1;
+			// gets the location that needs to be render of icon
+			Point2D location = foodBarOffsets.get(i);
+			if (location == null)
+				continue;
+
+			int x = (int)location.getX() + 0;
+			int y = (int)location.getY() + 0;
+
+			// location to normal food by default
+			int v = 3 * iconSize;
+			int u = iconStartOffset + 4 * iconSize;
+			int ub = iconStartOffset + 1 * iconSize;
 
+			// relocation to rotten food
 			if (useRottenTextures)
 			{
-				icon += 36;
-				background = 13;
+				u += 4 * iconSize;
+				ub += 12 * iconSize;
 			}
 
+			// relocation to half food
+			if (modifiedFood < (i + 1) * 2)
+				u += 1 * iconSize;
+
 			// very faint background
-			RenderSystem.color4f(1.0F, 1.0F, 1.0F, alpha * 0.1f);
-			mc.inGameHud.drawTexture(matrixStack, x, y, 16 + background * 9, 27, 9, 9);
+			RenderSystem.color4f(1.0F, 1.0F, 1.0F, alpha * 0.1F);
+			mc.inGameHud.drawTexture(matrixStack, x, y, ub, v, iconSize, iconSize);
 			RenderSystem.color4f(1.0F, 1.0F, 1.0F, alpha);
 
-			if (idx < foodLevel + hungerRestored)
-				mc.inGameHud.drawTexture(matrixStack, x, y, icon + 36, 27, 9, 9);
-			else if (idx == foodLevel + hungerRestored)
-				mc.inGameHud.drawTexture(matrixStack, x, y, icon + 45, 27, 9, 9);
+			mc.inGameHud.drawTexture(matrixStack, x, y, u, v, iconSize, iconSize);
 		}
+
+		disableAlpha(alpha);
+	}
+
+	public static void drawHealthOverlay(MatrixStack matrixStack, float health, float modifiedHealth, MinecraftClient mc, int offsetX, int offsetY, float alpha)
+	{
+		if (modifiedHealth <= health)
+			return;
+
+		enableAlpha(alpha);
+		mc.getTextureManager().bindTexture(Screen.GUI_ICONS_TEXTURE);
+
+		PlayerEntity player = mc.player;
+
+		int startHealthBars = (int)Math.floor(Math.ceil(health) / 2.0F);
+		int endHealthBars = (int)Math.ceil(modifiedHealth / 2.0F);
+
+		int iconStartOffset = 16;
+		int iconSize = 9;
+
+		for (int i = startHealthBars; i < endHealthBars; ++i)
+		{
+			// gets the location that needs to be render of icon
+			Point2D location = healthBarOffsets.get(i);
+			if (location == null)
+				continue;
+
+			int x = (int)location.getX() + offsetX;
+			int y = (int)location.getY() + offsetY;
+
+			// location to full heart icon by default
+			int v = 0 * iconSize;
+			int u = iconStartOffset + 4 * iconSize;
+			int ub = iconStartOffset + 1 * iconSize;
+
+			// relocation to half heart
+			if (modifiedHealth < (i + 1) * 2)
```

Comment:
This should be
```suggestion
			if (modifiedHealth < i * 2 + 1)
```
(this is what was causing the estimation to look wrong at 1:04 in the video; the estimated health was correct but the last heart wasn't being drawn correctly)